### PR TITLE
feat: #82/checkout middleware

### LINKED ISSUE #82: [Feature]: implement checkout middleware and fix payment error
### How the project can be improved?

Currently we do not have checkout middleware so client can go to customer and then click to move to payment directly and it will result in a wrong flow of order. Also We will have a transition to thank you even thought the process was not handled correctly.

image

### What are the acceptance criteria?

- [ ] Create checkout middleware that will allow user to move further in the checkout process only when the required data has been added
- [ ] Handle order payment error similarly to how it was handled when there was existing customer set for order.

### Additional information

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Baroshem**
## Description

Closes #82 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/theme/middleware/checkout.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      if (!canEnterPayment(activeCart)) {
+        app.context.redirect('/');
+      }
+      break;
+    case 'thank-you':
+      if (!canEnterThankYou(app.context)) {
+        app.context.redirect('/');
+      }
+      break;
+  }
```
Comment (Mateuszpietrusinski) [inline packages/theme/middleware/checkout.js:34]: Maybe better option is to have just if.
```suggestion
    if(currentPath === 'shipping' && !canEnterShipping(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'billing' && !canEnterBilling(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'payment' && !canEnterPayment(activeCart))  {
      app.context.redirect('/');
      return;
    }
     if(currentPath === 'thank-you' && !canEnterThankYou(app.context))  {
      app.context.redirect('/');
      return;
    }
  }
```

Also IMO good idea is to replace strings via const or enum values

File: packages/theme/pages/Checkout/Customer.vue [hunk 2]
```diff
@@ -140,6 +142,18 @@ export default {
       isFormSubmitted.value = true;
     };
 
+    onMounted(async () => {
```
Comment (Mateuszpietrusinski) [inline packages/theme/pages/Checkout/Customer.vue:145]: Maybe  its better to load that onSSR, this will verify quicker if user has access to this page

**Review by Mateuszpietrusinski — APPROVED** [review]
Some improvements, but LGTM

File: packages/theme/pages/Checkout/Customer.vue (hunk 2 above)
Comment (Baroshem) [reply packages/theme/pages/Checkout/Customer.vue:145]: I tried to put it in onSSRc lifecycle hook at first but it did not work. Now when you have already filled the customer step data and you are in other step (like shipping) when you come back to customer the data will be filled for you.

File: packages/theme/middleware/checkout.js (hunk 1 above)
Comment (Fifciu) [reply packages/theme/middleware/checkout.js:34]: const or enum values sound cool for me too

**Pull request merged**